tools/koch/builder.nim

```diff
@@ -0,0 +1,439 @@
+  for targetWithDep in dependencies.keys:
+    starters.excl targetWithDep
+
+  # Produce a topological ordering using Kahn's algorithm
+  while starters != {}:
+    let target = starters.pop()
+    result.add target
+
+    for targetWithDep, deps in dependencies.mpairs:
+      if target in deps:
+        deps.excl target
+
+        if deps == {}:
+          starters.incl targetWithDep
+
+  # Why is this an assert?
+  #
+  # The graph is constructed by declaring a node and its dependency at the same
+  # time, so it is impossible for a later node to be referenced by a previous
+  # one.
+  #
+  # If this happens, we messed up the code.
+  for deps in dependencies.values:
+    assert deps == {}:
+      "There is a cycle in the dependency graph"
+
+func sortTargets(b: var Builder): seq[Target] =
+  ## Perform a topological sort on builders' targets.
+  let sortedOrder = b.topologicalOrder
+
+  var
+    newNames: typeof(b.names)
+    newRecipes: typeof(b.recipes)
+    newDependencies: typeof(b.dependencies)
+
+  for idx, target in sortedOrder.pairs:
+    # Insert to new target seq according to the sorted order
+    newRecipes.add b.recipes[target]
+    newNames.add b.names[target]
+
+    # If the target has any dependencies
+    let deps = b.dependencies.getOrDefault(target)
+    if deps != {}:
+      for newDep in Target(0) ..< target:
+        let dep = sortedOrder[newDep]
+        # If dep was a dependency of the target
+        if dep in deps:
+          # Re-link it to the new ID
+          newDependencies[target].incl newDep
+
+  b.names = newNames
+  b.recipes = newRecipes
+  b.dependencies = newDependencies
+
+func name*(b: Builder, target: Target): string =
+  ## Returns the name of a given target
+  b.names[target]
+
+func addExecutable*(b: var Builder, name, src: string,
+                    extraFlags: varargs[string]): Executable =
+  ## Add a target for building a Nim executable. `name` is the name of the executable (sans extension).
+  let name =
+    if name.len == 0:
+      src.changeFileExt(ExeExt)
+    else:
+      name
+
+  # Construct a recipe for the executable
+  var recipe: TargetRecipe
+  recipe.instr = Build
+  recipe.a = Operand b.addInputs(src.atRoot(b.sourceRoot))
+
+  if extraFlags.len > 0:
+    recipe.instr = BuildWithArgs
+    # Append the parameters right after the input
+    discard b.addInputs(extraFlags)
+    # Record the length of the parameters
+    recipe.b = Operand extraFlags.len
+
+  result = Executable b.addTarget(name, recipe)
+
+func builtExe*(b: Builder, exe: Executable): string =
+  ## Returns the path to the built executable
+  # TODO: Move this to the build directory
+  b.sourceRoot / "bin" / b.name(exe).addFileExt(ExeExt)
+
+func cacheDirectory(b: Builder, exe: Executable): string =
+  ## Returns the path to the nimcache of the executable
+  b.buildDir / "nimcache" / b.name(exe)
+
+func addRun*(b: var Builder, name: string, exe: Executable,
```

Comment:
To get transactional behavior:
1. Immediately make a copy of b
2. Wrap the current body in a try
3. manipulate the copy instead of b in the try
4. in the last line of the try copy back the changes over from the copy into b
5. profit ... and by that I mean you probably want to make that a template